# Pandas. Домашнее задание

В домашнем задании есть обязательные и дополнительные задачи.

*Обязательные задачи* нужно сделать, чтобы проверить, что вы действительно усвоили материал модуля. Они обязательны к загрузке на проверку преподавателю.

*Дополнительные задачи* для тех, кто чувствует, что нужно ещё немного потренироваться и порешать задачки по работе с Pandas. Вы также можете загрузить их на проверку преподавателю и получить по ним обратную связь.

Удачи!



---



---



**Цели домашнего задания:**
1. Научиться работать с описательными статистиками набора данных.
2. Научиться выбирать данные по условиям.
3. Научиться объединять несколько датафреймов в один.


**Что входит в домашнее задание:**
1. *Обязательные задачи*:

17 задач с использованием:
- набора рекламных данных, показывающих, щелкнул ли конкретный пользователь интернета рекламу на веб-сайте компании или нет,
- дополнительного набора данных, описывающего пользователей, посещающих веб-сайт компании.
2. *Дополнительные задачи*:

3 задачи для отработки исследования данных.


**Критерии оценки:**
- программа выдаёт верный ответ на заданном наборе данных;
- код читабелен: переменным даны осмысленные названия, отступы и правила расстановки пробелов соблюдены.

**Как отправить задание на проверку**

Выполните задачи, сохраните изменения при помощи опции Save and Checkpoint из вкладки меню File или кнопки Save and Checkpoint на панели инструментов. Отправьте через форму на платформе итоговый файл Jupyter Notebook (в формате .ipynb) или ссылку на него.



---


---



**Обязательные задачи**

В течение всего домашнего задания вы будете работать с несколькими наборами данных:
1. Набор рекламных данных, показывающих, щёлкнул ли конкретный пользователь рекламу на сайте компании или нет. Состоит из трёх разных файлов: `advertising_1.csv`, `advertising_2.csv`, `advertising_3.csv`.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Daily Time Spent on Site**: Среднее ежедневное время нахождения пользователя на сайте в минутах.
- **Daily Internet Usage**: Среднее ежедневное время нахождения пользователя в интернете в минутах.
- **Ad Topic Line**: Заголовок рекламного объявления.
- **Clicked on Ad**: Кликнул пользователь на рекламу или нет.

Дополнительный к рекламным данным датасет, который описывает пользователей, посещающих сайт компании. Состоит из одного файла: users.csv.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Age**: Возраст пользователя.
- **Male**: Является ли пользователь мужчиной.
- **City**: Город пользователя.
- **Country**: Страна пользователя.
- **Area Income**: Средний доход в месте жительства пользователя.


***Задача 1***

Загрузите таблицу из файла `advertising_1.csv` и сохраните её в датафрейм `adv1_df`. Укажите в качестве индекса столбец `Number`. Выведите на экран первые десять строк.


In [1]:
# Ваш код здесь
import pandas as pd
import numpy as np
adv1_df = pd.read_csv('12.6 Данные для домашнего задания/advertising_1.csv', sep = ',', index_col='Number')
adv1_df.head(10)


Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
11                         88.91                208.36   
12                           NaN                131.76   
13                         74.53                   NaN   
14                         69.88                183.82   
15                         47.64                122.02   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
11                 Enhanced dedicated support              0  
12                   Reactive local challenge              1  
13             Configurable coherent function              0  
14         Mandatory homogeneous architecture              0  
15             Centralized neutral neural-net              1

***Задача 2***

Выведите размер датафрейма `adv1_df`.

In [2]:
# Ваш код здесь
print("Размер датафрейма `adv1_df`", adv1_df.shape, "строк:", adv1_df.shape[0], "столбцов:", adv1_df.shape[1])

Размер датафрейма `adv1_df` (500, 4) строк: 500 столбцов: 4


***Задача 3***

Выведите cреднее ежедневное время нахождения в интернете пользователя под номером 8.

In [3]:
# Ваш код здесь
print("Среднее ежедневное время нахождения в интернете пользователя под номером 8 = ",adv1_df['Daily Internet Usage'][8])

Среднее ежедневное время нахождения в интернете пользователя под номером 8 =  245.89


***Задача 4***

Загрузите таблицу из файла `advertising_2.csv` и сохраните её в датафрейм `adv2_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран строки для пользователей с номерами с 533-го по 536-й.

In [4]:
# Ваш код здесь
adv2_df = pd.read_csv('12.6 Данные для домашнего задания/advertising_2.csv', sep = ',', index_col='Number')
adv2_df.loc[533:536,]

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
533                        46.61                   NaN   
534                          NaN                162.03   
535                        65.10                118.10   
536                        53.44                   NaN   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
533             Ergonomic methodical encoding              1  
534         Compatible dedicated productivity              0  
535            Up-sized real-time methodology              1  
536     Up-sized next generation architecture              1

***Задача 5***

Выведите сводную таблицу описательных статистик для датафрейма `adv2_df`, вызвав необходимый метод на всём датафрейме. 

Глядя на вывод, напишите, столбец с каким названием не содержит незаполненных значений.

*Подсказка:*  В сводной таблице не содержится информации о количестве строк всего датафрейма, поэтому для сначала распечатайте общее количество строк датафрейма.

*Рекомендация:* Кроме изученного метода `describe` для описания датафрейма и получения информации о пропущенных значениях полезен метод [`info`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html).

In [5]:
# Ваш код здесь
print("Размер датафрейма `adv1_df`", adv2_df.shape, "строк:", adv2_df.shape[0], "столбцов:", adv2_df.shape[1])
print(adv2_df.shape)
print(adv2_df.describe())
print(adv2_df.info())

Размер датафрейма `adv1_df` (200, 4) строк: 200 столбцов: 4
(200, 4)
       Daily Time Spent on Site  Daily Internet Usage  Clicked on Ad
count                184.000000            184.000000     200.000000
mean                  66.494728            186.809674       0.420000
std                   15.414725             43.844580       0.494797
min                   32.910000            108.030000       0.000000
25%                   52.805000            148.955000       0.000000
50%                   71.150000            194.170000       0.000000
75%                   78.352500            225.372500       1.000000
max                   91.430000            259.760000       1.000000
<class 'pandas.core.frame.DataFrame'>
Int64Index: 200 entries, 530 to 739
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Daily Time Spent on Site  184 non-null    float64
 1   Daily Internet Usage      184 non-

In [6]:
# Ваш код здесь
print('Не содержит незаполненных значений столбец:',[col for col in adv2_df.columns if adv2_df[col].count() == adv2_df.shape[0]])

Не содержит незаполненных значений столбец: ['Ad Topic Line', 'Clicked on Ad']


***Задача 6***

Добавьте к датафрейму `adv1_df` в конец датафрейм `adv2_df` с помощью метода `append`. 

Сохраните объединенный датафрейм в новой переменной `adv12_df` и распечатайте его.

In [7]:
# Ваш код здесь
adv12_df = adv1_df.append(adv2_df)
adv12_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
735                        65.53                190.17   
736                        61.87                   NaN   
737                        83.16                   NaN   
738                        44.11                121.24   
739                        56.57                131.98   

                                    Ad Topic Line  Clicked on Ad  
Number                                                            
5              Cloned 5thgeneration orchestration              0  
6              Monitored national standardization              0  
8           Triple-buffered reciprocal time-frame              0  
9                   Robust logistical utilization              0  
10                Sharable client-driven software              0  
...                                           ...            ...  
735        Pre-emptive client-server installation              0  
736     Multi-channeled attitude-oriented toolset              0  
737                 Decentralized 24hour approach              0  
738                Organic next generation matrix              1  
739          Multi-channeled non-volatile website              1  

[700 rows x 4 columns]

In [8]:
adv12_df.shape

(700, 4)

***Задача 7***

Для объединённого датафрейма `adv12_df` распечатайте среднее время, проводимое пользователем на сайте компании.

In [9]:
# Ваш код здесь
adv12_df['Daily Time Spent on Site']

Number
5      68.95
6      80.23
8      74.15
9        NaN
10     59.99
       ...  
735    65.53
736    61.87
737    83.16
738    44.11
739    56.57
Name: Daily Time Spent on Site, Length: 700, dtype: float64

***Задача 8***

Из объединённого датафрейма `adv12_df` выберите и распечатайте строки только с теми пользователями, для которых неизвестно среднее ежедневное время, проводимое пользователем на сайте компании.

In [10]:
# Ваш код здесь
adv12_df[adv12_df['Daily Time Spent on Site'].isna()]

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
9                            NaN                225.58   
12                           NaN                131.76   
19                           NaN                143.56   
27                           NaN                164.83   
34                           NaN                209.82   
...                          ...                   ...   
672                          NaN                235.97   
674                          NaN                126.44   
685                          NaN                200.71   
690                          NaN                226.69   
731                          NaN                225.05   

                                  Ad Topic Line  Clicked on Ad  
Number                                                          
9                 Robust logistical utilization              0  
12                     Reactive local challenge              1  
19                Grass-roots coherent extranet              1  
27        Mandatory disintermediate utilization              1  
34      Business-focused value-added definition              0  
...                                         ...            ...  
672            Customizable value-added project              0  
674                Reactive impactful challenge              1  
685               Focused multi-state workforce              0  
690               Total coherent superstructure              0  
731             Multi-tiered heuristic strategy              0  

[66 rows x 4 columns]

***Задача 9***

Загрузите таблицу из файла `advertising_3.csv` и сохраните её в датафрейм `adv3_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран столбцы `Ad Topic Line` и `Clicked on Ad`.

In [11]:
# Ваш код здесь
adv3_df = pd.read_csv('12.6 Данные для домашнего задания/advertising_3.csv', sep = ',', index_col='Number')
adv3_df[['Ad Topic Line','Clicked on Ad']]


Ad Topic Line  Clicked on Ad
Number                                                             
740                 Distributed bifurcated challenge              0
741       Customizable zero-defect Internet solution              0
742     Self-enabling zero administration neural-net              1
744           Open-architected system-worthy ability              0
745               Quality-focused maximized extranet              1
...                                              ...            ...
999                    Front-line bifurcated ability              1
1000                   Fundamental modular algorithm              1
1001                 Grass-roots cohesive monitoring              1
1002                    Expanded intangible solution              1
1003            Proactive bandwidth-monitored policy              0

[251 rows x 2 columns]

***Задача 10***

Добавьте к датафрейму `adv12_df` в конец датафрейм `adv3_df` с помощью метода `concat`. Явно пропишите значение параметра `axis`.

Сохраните объединённый датафрейм в новой переменной `adv123_df` и распечатайте его.

In [12]:
# Ваш код здесь
adv123_df = pd.concat([adv12_df, adv3_df], axis=0)
adv123_df.tail()

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
999                        43.70                173.01   
1000                       72.97                208.58   
1001                       51.30                   NaN   
1002                       51.63                120.37   
1003                       55.55                187.95   

                               Ad Topic Line  Clicked on Ad  
Number                                                       
999            Front-line bifurcated ability              1  
1000           Fundamental modular algorithm              1  
1001         Grass-roots cohesive monitoring              1  
1002            Expanded intangible solution              1  
1003    Proactive bandwidth-monitored policy              0

In [13]:
adv123_df.shape

(951, 4)

***Задача 11***

Для объединённого датафрейма `adv123_df` выберите только те строки, в которых пользователи кликнули на рекламное сообщение. Сохраните полученный результат в датафрейм `success_adv_df`, распечатайте размер нового датафрейма.

In [14]:
# Ваш код здесь
success_adv_df = adv123_df[adv123_df['Clicked on Ad'] == 1]
success_adv_df.tail()

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
997                        69.17                123.62   
999                        43.70                173.01   
1000                       72.97                208.58   
1001                       51.30                   NaN   
1002                       51.63                120.37   

                          Ad Topic Line  Clicked on Ad  
Number                                                  
997       Extended leadingedge solution              1  
999       Front-line bifurcated ability              1  
1000      Fundamental modular algorithm              1  
1001    Grass-roots cohesive monitoring              1  
1002       Expanded intangible solution              1

In [15]:
success_adv_df.shape

(475, 4)

***Задача 12***

Загрузите таблицу с данными пользователей из файла `users.csv` и сохраните её в датафрейм `users_df`. Индекс не указывайте.

Распечатайте сам датафрейм и сводную таблицу со статистиками для него.

In [295]:
# Ваш код здесь
user_df = pd.read_csv('12.6 Данные для домашнего задания/users.csv', sep = ',')
user_df.head(10)

Number   Age  Male              City     Country  Area Income
0       5   NaN   0.0       Wrightburgh     Tunisia     61833.90
1       6  31.0   NaN         West Jodi       Nauru     68441.85
2       7  26.0   0.0          Davidton  San Marino     59785.94
3       8  29.0   1.0    West Terrifurt       Italy     54806.18
4       9   NaN   0.0      South Manuel     Iceland     73889.99
5      10  23.0   1.0               NaN      Norway     59761.56
6      11  33.0   0.0       Brandonstad     Myanmar     53852.85
7      12  48.0   1.0  Port Jefferybury   Australia     24593.33
8      14  20.0   NaN        Ramirezton       Ghana     55642.32
9      15  49.0   NaN   West Brandonton       Qatar     45632.51

In [17]:
user_df.describe()

Number         Age        Male   Area Income
count   980.000000  860.000000  868.000000    903.000000
mean    504.485714   36.073256    0.500000  55010.631938
std     287.956480    8.795992    0.500288  13423.710892
min       5.000000   19.000000    0.000000  13996.500000
25%     255.750000   29.000000    0.000000  46947.570000
50%     503.500000   35.000000    0.500000  57032.360000
75%     753.250000   42.000000    1.000000  65537.990000
max    1004.000000   61.000000    1.000000  79484.800000

In [18]:
user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 980 entries, 0 to 979
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Number       980 non-null    int64  
 1   Age          860 non-null    float64
 2   Male         868 non-null    float64
 3   City         887 non-null    object 
 4   Country      888 non-null    object 
 5   Area Income  903 non-null    float64
dtypes: float64(3), int64(1), object(2)
memory usage: 46.1+ KB


***Задача 13***

Глядя на статистики `users_df`, укажите значения, удовлетворяющие следующим условиям:
1. Возраст самого молодого пользователя сайта: *19.000000*
2. Максимальный средний доход в месте жительства пользователя сайта: *79484.800000*
3. Средний возраст пользователя сайта: *55010.631938*
4. Количество заполненных значений для колонки, содержащей информацию о том, является пользователь мужчиной или нет: *868*
5. Минимальный номер пользователя: *5*

В этом задании не требуется выбирать по условиям и распечатывать значения, достаточно внимательно посмотреть на статистики и скопировать нужные значения согласно вопросам.

***Задача 14***

К полученному датафрейму с успешными рекламными сообщениями `success_adv_df` добавьте справа информацию из датафрейма с пользователями `users_df` и сохраните в датафрейм `success_full_df`. В полученном датафрейме должны остаться только те пользователи, номера которых есть в обоих датафреймах. Для этого воспользуйтесь методом `merge` и пропишите явно верные параметры.

Затем распечатайте датафрейм.

*Важно:*

Объединять датафреймы с помощью `merge` можно и по индексам, и по столбцам с разными названиями. 
Так как общие значения содержатся в столбце `Number`, но в датафрейме `users_df` данный столбец является обычным столбцом, а в `success_adv_df` является индексом, перед тем как объединять, переведите столбец `Number` в `success_adv_df` в обычный столбец.

Для этого перед объединением пропишите:
```
success_adv_df = success_adv_df.reset_index()
```

[Подробнее о reset_index](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html)

In [19]:
# Ваш код здесь
success_adv_df = success_adv_df.reset_index()
success_full_df = success_adv_df.merge(user_df, left_on='Number', right_on='Number')
success_full_df.shape

(464, 10)

In [22]:
success_full_df.head(10)

Number  Daily Time Spent on Site  Daily Internet Usage  \
0      12                       NaN                131.76   
1      15                     47.64                122.02   
2      17                     69.57                113.12   
3      19                       NaN                143.56   
4      20                     63.45                140.64   
5      21                     55.39                   NaN   
6      24                     74.58                135.51   
7      27                       NaN                164.83   
8      29                     41.39                   NaN   
9      31                     48.53                134.14   

                           Ad Topic Line  Clicked on Ad   Age  Male  \
0               Reactive local challenge              1  48.0   1.0   
1         Centralized neutral neural-net              1  49.0   NaN   
2  Centralized content-based focus group              1  48.0   1.0   
3          Grass-roots coherent extranet              1  33.0   0.0   
4     Persistent demand-driven interface              1  23.0   NaN   
5     Customizable multi-tasking website              1   NaN   0.0   
6             Advanced 24/7 productivity              1  40.0   1.0   
7  Mandatory disintermediate utilization              1  52.0   0.0   
8          Exclusive neutral parallelism              1  41.0   0.0   
9     Ameliorated client-driven forecast              1   NaN   1.0   

               City                Country  Area Income  
0  Port Jefferybury              Australia     24593.33  
1   West Brandonton                  Qatar     45632.51  
2    West Katiefurt                  Egypt     51636.92  
3               NaN               Barbados     30976.00  
4               NaN                  Spain     52182.23  
5    West Dylanberg  Palestinian Territory     23936.86  
6        Millertown     Russian Federation     23821.72  
7        South John                Burundi     32635.70  
8     Harperborough                Tokelau     68962.32  
9   West Jeremyside                 Tuvalu     38067.08

***Задача 15***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, отобразите сводную таблицу со статистиками для строковых колонок (`Ad Topic Line`, `City`, `Country`).


In [24]:
# Ваш код здесь
success_full_df[['Ad Topic Line', 'City', 'Country']].describe()

Ad Topic Line        City   Country
count                        464         420       418
unique                       464         412       203
top     Reactive local challenge  Lake James  Ethiopia
freq                           1           2         7

In [25]:
success_full_df[['Ad Topic Line', 'City', 'Country']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 464 entries, 0 to 463
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Ad Topic Line  464 non-null    object
 1   City           420 non-null    object
 2   Country        418 non-null    object
dtypes: object(3)
memory usage: 14.5+ KB


***Задача 16***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите страну и количество вхождений этой страны в датафрейм.


In [160]:
# Ваш код здесь
count_Country = pd.value_counts(success_full_df['Country']).to_dict()
maxCountryUsers = [k for k, v in count_Country.items() if v == max(count_Country.values())]
count_Country


{'Ethiopia': 7,
 'Australia': 6,
 'Turkey': 6,
 'Liberia': 6,
 'Senegal': 5,
 'Peru': 5,
 'South Africa': 5,
 'Hong Kong': 4,
 'Mongolia': 4,
 'El Salvador': 4,
 'Uganda': 4,
 'Afghanistan': 4,
 'Zimbabwe': 4,
 'Lebanon': 4,
 'Christmas Island': 4,
 'Cyprus': 4,
 'Indonesia': 4,
 'Samoa': 4,
 'Antigua and Barbuda': 4,
 'Hungary': 4,
 'Taiwan': 4,
 'France': 4,
 'Bulgaria': 4,
 'Czech Republic': 4,
 'Anguilla': 3,
 'Jersey': 3,
 'Liechtenstein': 3,
 'China': 3,
 'Belarus': 3,
 'Bahamas': 3,
 'Saint Vincent and the Grenadines': 3,
 'Eritrea': 3,
 'Cuba': 3,
 'Guyana': 3,
 'Mexico': 3,
 'Philippines': 3,
 'Venezuela': 3,
 'Fiji': 3,
 'Equatorial Guinea': 3,
 'Mayotte': 3,
 'Tuvalu': 3,
 'Papua New Guinea': 3,
 'Bosnia and Herzegovina': 3,
 'Latvia': 3,
 'Chile': 3,
 'Albania': 3,
 'Spain': 3,
 'Micronesia': 3,
 'Saint Pierre and Miquelon': 3,
 'Cayman Islands': 3,
 'Macao': 3,
 'Turks and Caicos Islands': 3,
 'Egypt': 3,
 'French Guiana': 3,
 'American Samoa': 3,
 'Algeria': 3,
 'Luxembou

***Задача 17***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных.


In [161]:
# Ваш код здесь
success_full_df[success_full_df['Country'] == maxCountryUsers[0]]

Number  Daily Time Spent on Site  Daily Internet Usage  \
245     525                     46.61                136.18   
246     526                     63.26                120.46   
278     606                     52.13                118.27   
297     650                       NaN                137.97   
299     652                     46.43                137.20   
417     919                     41.86                128.62   
423     930                       NaN                   NaN   

                                       Ad Topic Line  Clicked on Ad   Age  \
245              Progressive intermediate throughput              1  42.0   
246                    Customizable holistic archive              1  29.0   
278                 Open-source even-keeled database              1  50.0   
297      Seamless bandwidth-monitored knowledge user              1  28.0   
299  Reverse-engineered background Graphic Interface              1  28.0   
417                      Open-source stable paradigm              1  39.0   
423           Front-line fresh-thinking installation              1  45.0   

     Male               City   Country  Area Income  
245   0.0      Catherinefort  Ethiopia     65856.74  
246   1.0         East Donna  Ethiopia     54787.37  
278   1.0  West Raymondmouth  Ethiopia     40926.93  
297   1.0          Markhaven  Ethiopia     39723.97  
299   1.0          Brianland  Ethiopia     53922.43  
417   1.0        Hawkinsbury  Ethiopia     53041.77  
423   0.0     Huffmanchester  Ethiopia     43241.88

***Задача 18***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных, они проводят в интернете в среднем больше 120 минут и их возраст меньше 30 лет. 
Cохраните получившиеся строки в новый датафрейм `filtered_df`.


In [187]:
# Ваш код здесь
filtered_df = success_full_df.loc[(success_full_df['Country'] == maxCountryUsers[0])
                                  & (success_full_df['Daily Internet Usage'] > 120.0)
                                  & (success_full_df['Age'] < 30.0)
                                 ]
filtered_df

Number  Daily Time Spent on Site  Daily Internet Usage  \
246     526                     63.26                120.46   
297     650                       NaN                137.97   
299     652                     46.43                137.20   

                                       Ad Topic Line  Clicked on Ad   Age  \
246                    Customizable holistic archive              1  29.0   
297      Seamless bandwidth-monitored knowledge user              1  28.0   
299  Reverse-engineered background Graphic Interface              1  28.0   

     Male        City   Country  Area Income  
246   1.0  East Donna  Ethiopia     54787.37  
297   1.0   Markhaven  Ethiopia     39723.97  
299   1.0   Brianland  Ethiopia     53922.43

***Задача 19***

Для получившегося датафрейма `filtered_df` выведите строки с данными пользователей, для которых либо не указано среднее время, проведённое на сайте, либо указанное время больше 55 минут. Cохраните получившиеся строки в новый датафрейм `filtered_df2`.


In [189]:
# Ваш код здесь
filtered_df2 = filtered_df.loc[(filtered_df['Daily Time Spent on Site'].isna())
    | (filtered_df['Daily Time Spent on Site'] > 55.0)
]
filtered_df2

Number  Daily Time Spent on Site  Daily Internet Usage  \
246     526                     63.26                120.46   
297     650                       NaN                137.97   

                                   Ad Topic Line  Clicked on Ad   Age  Male  \
246                Customizable holistic archive              1  29.0   1.0   
297  Seamless bandwidth-monitored knowledge user              1  28.0   1.0   

           City   Country  Area Income  
246  East Donna  Ethiopia     54787.37  
297   Markhaven  Ethiopia     39723.97

**Дополнительные задачи**

***Задача 20***

Для датафрейма `adv123_df` выведите строки с данными, заголовок рекламного сообщения в которых включает слово `model`.

Для этого вам понадобится метод [`contains`](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html).

Cохраните получившиеся строки в новый датафрейм `model_df`.


In [293]:
# Ваш код здесь
model_df = adv123_df[adv123_df['Ad Topic Line'].str.contains('model')]
model_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
62                         44.98                129.31   
119                        69.00                   NaN   
202                        78.57                   NaN   
211                        76.90                212.67   
314                          NaN                161.16   
350                        65.77                   NaN   
501                        81.10                216.49   
538                        91.43                209.91   
629                        78.17                192.27   
712                        81.90                225.47   
823                        81.98                212.88   
995                        35.79                165.62   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
62                   Organized global model              1  
119                 Virtual composite model              0  
202           Exclusive client-driven model              0  
211           Synchronized multimedia model              0  
314              Extended interactive model              1  
350     Function-based fault-tolerant model              0  
501                Front-line dynamic model              0  
538                Organic motivating model              0  
629                  Automated mobile model              0  
712     Multi-channeled 3rdgeneration model              0  
823                Up-sized maximized model              0  
995          Enterprise-wide tangible model              1

***Задача 21***

Для получившегося датафрейма `model_df` выведите строки с данными пользователей, которые проводят на сайте больше среднего времени по всем пользователям в данном датафрейме. Cохраните получившиеся строки в новый датафрейм `model_popular_df`.


In [324]:
# Ваш код здесь
user1_df = user_df.set_index('Number')
model_popular_df = user1_df.merge(model_df[model_df['Daily Time Spent on Site'] > model_df['Daily Time Spent on Site'].describe()['mean']], left_index=True, right_index=True)# model_df['Daily Time Spent on Site'].describe()['mean']
model_popular_df

Age  Male             City           Country  Area Income  \
Number                                                               
202      NaN   1.0       East Henry           Liberia     64045.93   
211     28.0   0.0      Murphymouth           Croatia     66107.84   
501     29.0   1.0       Port Stacy       El Salvador     61172.07   
538     39.0   1.0  West Carmenfurt  French Polynesia     46964.11   
629     27.0   1.0      Portermouth           Burundi     65227.79   
712     41.0   0.0       Garciatown         Mauritius     66773.83   
823     34.0   0.0     Patrickmouth      Saint Martin     67432.49   

        Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
202                        78.57                   NaN   
211                        76.90                212.67   
501                        81.10                216.49   
538                        91.43                209.91   
629                        78.17                192.27   
712                        81.90                225.47   
823                        81.98                212.88   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
202           Exclusive client-driven model              0  
211           Synchronized multimedia model              0  
501                Front-line dynamic model              0  
538                Organic motivating model              0  
629                  Automated mobile model              0  
712     Multi-channeled 3rdgeneration model              0  
823                Up-sized maximized model              0

In [327]:
model_df['Daily Time Spent on Site'].describe()['mean']

71.41727272727272

***Задача 22***

Для объединённого датафрейма `model_popular_df` выведите строки с данными пользователей, номера которых равны 501, 712 и 823. 

Для этого вам понадобится метод [`isin`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isin.html).

*Подсказка:* Не забудьте, что номера пользователей находятся не в обычной колонке, а в колонке, являющейся индексом, поэтому вам необходимо получить значения индекса, а не обычной колонки.

In [325]:
# Ваш код здесь
model_popular_df.loc[[501,712,823]]

Age  Male          City       Country  Area Income  \
Number                                                        
501     29.0   1.0    Port Stacy   El Salvador     61172.07   
712     41.0   0.0    Garciatown     Mauritius     66773.83   
823     34.0   0.0  Patrickmouth  Saint Martin     67432.49   

        Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
501                        81.10                216.49   
712                        81.90                225.47   
823                        81.98                212.88   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
501                Front-line dynamic model              0  
712     Multi-channeled 3rdgeneration model              0  
823                Up-sized maximized model              0